### PR TITLE
[cxx] Port profilers to C++. @monojenkins skip

### PR DESCRIPTION
**Pull request by jaykrell**
@monojenkins skip
 [cxx] Port profilers to C++.
    This alleviates need for externC in glib.

### CODE REVIEW

File: mono/profiler/log.c [hunk 1]
```diff
@@ -3301,7 +3313,8 @@ handle_writer_queue_entry (void)
 		if (!entry->methods)
 			goto no_methods;
 
-		gboolean wrote_methods = FALSE;
+		gboolean wrote_methods;
+		wrote_methods = FALSE;
```
Comment (alexrp) [inline mono/profiler/log.c]: These initialization changes make the code rather ugly. Why are they necessary?

Comment (jaykrell) [reply mono/profiler/log.c]: Cannot goto around initialization in C++.

Comment (jaykrell) [reply mono/profiler/log.c]: https://github.com/mono/mono/issues/10319#issuecomment-416873012

**Comment by jaykrell** [comment]
@alexrp @lambdageek we have decided to use extern "C" more in the runtime/glib.h, to workaround these non-core uses of non-public not-even-internal APIs, so this PR can be considered probably very optional this point. If you are ok with it, ok. If you don't want, ok too.

**Comment by jaykrell** [comment]
If we do switch to C++, we'll probably want to take this PR, so that we can use C++ freely in the internal mono headers that the profiler includes.

I'm trying C++ in handle.h/handle-decl.h but I have to keep C working because of the profiler.

Or move whatever profiler uses to other/public headers. Public headers have to remain C.

**Comment by kumpera** [comment]
If the provider is using internal headers, it’s fair game to convert them to c++.

**Comment by kumpera** [comment]
I mean profilers

**Comment by jaykrell** [comment]
Reopen and approve?

**Comment by jaykrell** [comment]
@alexrp?

File: mono/profiler/Makefile.am [hunk 2]
```diff
@@ -111,39 +112,51 @@ endif
...
 endif
 
-mprof_report_SOURCES = mprof-report.c
-mprof_report_LDADD = $(glib_libs) $(zlib_dep)
+mprof_report_SOURCES =
+mprof_report_LDADD = libmprof_report_a-mprof-report.$(OBJEXT) $(glib_libs) $(zlib_dep)
 mprof_report_LDFLAGS = -no-undefined
 
+libmprof_report_a_SOURCES = mprof-report.c
+libmprof_report_a_CFLAGS = @CXX_ADD_CFLAGS@
```
Comment (alexrp) [inline mono/profiler/Makefile.am]: It's not immediately obvious why this needs to be a static library just to change its `CFLAGS`.

File: mono/profiler/log.c [hunk 3]
```diff
@@ -16,7 +16,18 @@
 #include <mono/metadata/class-internals.h>
 #include <mono/metadata/debug-helpers.h>
 #include <mono/metadata/loader.h>
+#if 0 //FIXME
 #include <mono/metadata/loader-internals.h>
+#else
+#ifdef __cplusplus
+template <typename T>
+inline void
+mono_add_internal_call (const char *name, T method)
```
Comment (alexrp) [inline mono/profiler/log.c]: Just modify the `ADD_ICALL` macro to do the cast; no need for a template hack IMO.

Comment (alexrp) [reply mono/profiler/log.c]: And we are certain that, on the targets where we care about runtime size, those templates don't result in a bunch of pointless instantiations increasing the size of the final library?

File: mono/profiler/Makefile.am (hunk 2 above)
Comment (alexrp) [reply mono/profiler/Makefile.am]: Is this explained in some other place in our makefiles? If so, a comment here pointing to that explanation would be good.

File: mono/profiler/log.c (hunk 3 above)
Comment (jaykrell) [reply mono/profiler/log.c]: Yes. It is a one line inline function, that casts and calls the original.

If we templatized “deeply”, and didn’t inline, most?all systems combine identical functions (Microsoft linker got this feature circa 1996 and it isn’t alone.)

**Comment by jaykrell** [comment]
Reopen. Otherwise github not seeing updates.

**Comment by jaykrell** [comment]
Github would not let me reopen, so I opened another. https://github.com/mono/mono/pull/10886

**Pull request closed without merging**